src/main/java/io/pravega/connectors/flink/source/enumerator/PravegaSplitEnumerator.java

```diff
@@ -0,0 +1,206 @@
+            String scope,
+            String readerGroupName,
+            ClientConfig clientConfig,
+            ReaderGroupConfig readerGroupConfig,
+            Checkpoint checkpoint) {
+        this.enumContext = context;
+        this.scope = scope;
+        this.readerGroupName = readerGroupName;
+        this.clientConfig = clientConfig;
+        this.readerGroupConfig = readerGroupConfig;
+        this.splits = new ArrayList<>();
+        for (int i = 0; i < enumContext.currentParallelism(); i++) {
+            splits.add(new PravegaSplit(readerGroupName, i));
+        }
+        this.checkpoint = checkpoint;
+        this.scheduledExecutorService = Executors.newScheduledThreadPool(DEFAULT_CHECKPOINT_THREAD_POOL_SIZE);
+    }
+
+    // initiate reader group manager, reader group and reset the group to the checkpoint position if checkpoint isn't null
+    @Override
+    public void start() {
+        LOG.info("Starting the PravegaSplitEnumerator for reader group: {}/{}.", this.scope, this.readerGroupName);
+
+        if (this.readerGroupManager == null) {
+            this.readerGroupManager = ReaderGroupManager.withScope(scope, clientConfig);
+        }
+        if (this.readerGroup == null) {
+            this.readerGroupManager.createReaderGroup(this.readerGroupName, readerGroupConfig);
+            this.readerGroup = this.readerGroupManager.getReaderGroup(this.readerGroupName);
+        }
+        if (this.checkpoint != null) {
+            LOG.info("Recover from checkpoint: {}", checkpoint.getName());
+            this.readerGroup.resetReaderGroup(ReaderGroupConfig
+                    .builder()
+                    .maxOutstandingCheckpointRequest(this.readerGroupConfig.getMaxOutstandingCheckpointRequest())
+                    .groupRefreshTimeMillis(this.readerGroupConfig.getGroupRefreshTimeMillis())
+                    .disableAutomaticCheckpoints()
+                    .startFromCheckpoint(this.checkpoint)
+                    .build());
+        }
+    }
+
+    @Override
+    public void handleSplitRequest(int subtaskId, @Nullable String requesterHostname) {
+        // the Pravega source pushes splits eagerly, rather than act upon split requests
+    }
+
+    // check with the current parallelism and update the assignment (always one-to-one mapping)
+    @Override
+    public void addReader(int subtaskId) {
+        if (enumContext.registeredReaders().size() == enumContext.currentParallelism() && splits.size() > 0) {
+            LOG.debug("Adding reader {} to PravegaSplitEnumerator for reader group: {}/{}.",
+                    subtaskId,
+                    scope,
+                    readerGroupName);
+            int numReaders = enumContext.registeredReaders().size();
+            Map<Integer, List<PravegaSplit>> assignment = new HashMap<>();
+            for (int i = 0; i < numReaders; i++) {
+                assignment.put(i, Collections.singletonList(splits.get(i)));
```

Comment:
Could you please document as a comment how this assignment is being performed? The `subtaskId` parameter isn't used for the assignment, and we seem to be assigning `i` to `i` here, which seems to match the desired mapping of subtask ID to one split. I'd like to see some explanation as of why the assignment is being performed this way.

Reply:
I have changed the implementation here to perform the assignment once this method is called by assigning the Pravega Split with incoming `subtaskId` to Source Reader to make it more concise and more readable. Please help rereview it.

Also let me explain the original design briefly. It was intended to assign all splits at a time when all readers are registered. Here it won't go into the `if` statement until the number of register readers equals `currentParallelism`. Then we can perform the assignment in the `for` loop because the numbering of `subtaskId` starts from 0 and goes up to parallelism-1 according to [this](https://nightlies.apache.org/flink/flink-docs-stable/api/java/org/apache/flink/api/common/functions/RuntimeContext.html#getIndexOfThisSubtask--), which can map to the index `i`.